### PR TITLE
Check blob cache before getting blob file reader 

### PR DESCRIPTION
**Pull request by Connor1996**
Close https://github.com/tikv/titan/issues/140

Move blob cache query to before getting blob file reader, then we save the two mutex lock and other overheads for a cache hit query.

### CODE REVIEW

File: src/blob_storage.cc [hunk 1]
```diff
@@ -6,24 +6,71 @@
...
+  return cache_key;
+}
+
+Status BlobStorage::TryGetBlobCache(const std::string& cache_key,
+                                    BlobRecord* record, PinnableSlice* value,
+                                    bool* cache_hit) {
+  Status s;
+  Cache::Handle* cache_handle = blob_cache_->Lookup(cache_key);
+  if (cache_handle) {
+    *cache_hit = true;
```
Comment (tonyxuqqi) [inline src/blob_storage.cc:22]: `*cache_hit = true;` should be set only when s is ok.

File: src/db_iter.h [hunk 2]
```diff
@@ -154,7 +164,8 @@ class TitanDBIterator : public Iterator {
     }
 
     buffer_.Reset();
-    s = it->second->Get(options_, index.blob_handle, &record_, &buffer_);
+    OwnedSlice blob;
+    s = it->second->Get(options_, index.blob_handle, &record_, &blob);
```
Comment (tonyxuqqi) [inline src/db_iter.h:168]: Ideally when blob_cache is not enabled, then we can still reuse the buffer_ instead of have a new OwnedSlice.

**Review by tonyxuqqi — APPROVED** [review]
some minor comments

File: src/blob_storage.cc [hunk 3]
```diff
@@ -6,24 +6,71 @@
...
+  OwnedSlice blob;
+  Status s = file_cache_->Get(options, index.file_number, index.blob_handle,
+                              record, &blob);
+  if (!s.ok()) {
+    return s;
+  }
+
+  if (blob_cache_ && options.fill_cache) {
+    Cache::Handle* cache_handle = nullptr;
+    auto cache_value = new OwnedSlice(std::move(blob));
```
Comment (v01dstar) [inline src/blob_storage.cc:58]: Maybe using `blob.release()`?

File: src/db_impl.cc [hunk 4]
```diff
@@ -1486,5 +1486,16 @@ void TitanDBImpl::DumpStats() {
   log_buffer.FlushBufferToLog();
 }
 
+std::string TitanDBImpl::GenerateCachePrefix() {
+  std::string cache_prefix;
+  // Unlike block-based table, which have per-table cache prefix, we use
+  // the same cache prefix for all blob files for blob cache.
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc:1492]: To differentiate with block cache

File: src/blob_storage.cc (hunk 3 above)
Comment (Connor1996) [reply src/blob_storage.cc:58]: it returns `char *` which is not what OwnedSlicce wants

Comment (v01dstar) [reply src/blob_storage.cc:58]: I mean we can pass char* to `blob_cache_->Insert()`?

Comment (Connor1996) [reply src/blob_storage.cc:58]: We can't because we need a deleter `char*` which is implemented by `OwnedSlice`. The other place does the same way

**Comment by Connor1996** [comment]
PTAL again @v01dstar

**Review by v01dstar — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Pull request merged**